src/libasr/pass/intrinsic_array_function_registry.h

```
@@ -281,46 +284,84 @@ static inline ASR::expr_t *eval_ArrIntrinsic(Allocator & al,
     } else {
         return nullptr;
     }
+    ASR::ArrayConstant_t *mask = nullptr;
+    ASR::expr_t* dim = b.i32(1);
+    if (args[1]) {
+        if (is_integer(*ASRUtils::expr_type(args[1]))){
+            dim = args[1];
+            if (args[2]) {
+                if (ASR::is_a<ASR::ArrayConstant_t>(*args[2])) {
+                    mask = ASR::down_cast<ASR::ArrayConstant_t>(args[2]);
+                } else if (ASR::is_a<ASR::LogicalConstant_t>(*args[2])) {
+                    std::vector<ASR::expr_t*> mask_values(size, args[2]);
+                    ASR::expr_t *arr = b.ArrayConstant(mask_values, logical, false);
+                    mask = ASR::down_cast<ASR::ArrayConstant_t>(arr);
+                } else {
+                    return nullptr;
+                }
+            }
+        } else {
+            if (ASR::is_a<ASR::ArrayConstant_t>(*args[1])) {
+                mask = ASR::down_cast<ASR::ArrayConstant_t>(args[1]);
+            } else if (ASR::is_a<ASR::LogicalConstant_t>(*args[1])) {
+                std::vector<ASR::expr_t*> mask_values(size, args[1]);
+                ASR::expr_t *arr = b.ArrayConstant(mask_values, logical, false);
+                mask = ASR::down_cast<ASR::ArrayConstant_t>(arr);
+            } else {
+                return nullptr;
+            }
+        }
+    } else {
+        std::vector<ASR::expr_t*> mask_values(size, b.bool_t(true, logical));
+        ASR::expr_t *arr = b.ArrayConstant(mask_values, logical, false);
+        mask = ASR::down_cast<ASR::ArrayConstant_t>(arr);
```

One thing to keep in mind is that this `else` branch here is the most commonly used, since `mask` is rarely used. So it slows down compilation in the default case, as you need to create this array, and then the "if(mask)` logic in the actual summation.

This can be improved by distinguishing the two cases: mask / no mask, and handling the "no mask" branch as before.

---

aah yes, I missed it. Thank you for posting, I'll take care while reviewing other PRs.